packages/node-resolve/src/index.js

```diff
@@ -71,201 +72,220 @@ export function nodeResolve(opts = {}) {
-
-      const warn = (...args) => this.warn(...args);
-      const isRequire =
-        opts && opts.custom && opts.custom['node-resolve'] && opts.custom['node-resolve'].isRequire;
-      const exportConditions = isRequire ? conditionsCjs : conditionsEsm;
-
-      const resolvedWithoutBuiltins = await resolveImportSpecifiers({
-        importer,
-        importSpecifierList,
-        exportConditions,
-        warn,
-        packageInfoCache,
-        extensions,
-        mainFields,
-        preserveSymlinks,
-        useBrowserOverrides,
-        baseDir,
-        moduleDirectories
-      });
-
-      const resolved =
-        importeeIsBuiltin && preferBuiltins
-          ? {
-              packageInfo: undefined,
-              hasModuleSideEffects: () => null,
-              hasPackageEntry: true,
-              packageBrowserField: false
-            }
-          : resolvedWithoutBuiltins;
-      if (!resolved) {
-        return null;
-      }
+      ({ preserveSymlinks } = options);
+    },
 
-      const { packageInfo, hasModuleSideEffects, hasPackageEntry, packageBrowserField } = resolved;
-      let { location } = resolved;
-      if (packageBrowserField) {
-        if (Object.prototype.hasOwnProperty.call(packageBrowserField, location)) {
-          if (!packageBrowserField[location]) {
-            browserMapCache.set(location, packageBrowserField);
-            return ES6_BROWSER_EMPTY;
-          }
-          location = packageBrowserField[location];
-        }
-        browserMapCache.set(location, packageBrowserField);
-      }
+    generateBundle() {
+      readCachedFile.clear();
+      isFileCached.clear();
+      isDirCached.clear();
+    },
 
-      if (hasPackageEntry && !preserveSymlinks) {
-        const fileExists = await exists(location);
-        if (fileExists) {
-          location = await realpath(location);
-        }
+    async resolveId(importee, importer, opts) {
+      if (importee === ES6_BROWSER_EMPTY) {
+        return importee;
       }
+      // ignore IDs with null character, these belong to other plugins
+      if (/\0/.test(importee)) return null;
 
-      idToPackageInfo.set(location, packageInfo);
+      if (/\0/.test(importer)) {
+        importer = undefined;
+      }
 
-      if (hasPackageEntry) {
-        if (importeeIsBuiltin && preferBuiltins) {
-          if (!isPreferBuiltinsSet && resolvedWithoutBuiltins && resolved !== importee) {
-            this.warn(
-              `preferring built-in module '${importee}' over local alternative at '${resolvedWithoutBuiltins.location}', pass 'preferBuiltins: false' to disable this behavior or 'preferBuiltins: true' to disable this warning`
-            );
+      const resolved = await doResolveId(this, importee, importer, opts);
+      if (resolved && !inNestedResolve) {
+        // if a plugin invoked by `this.resolve` also does `this.resolve`
+        // we need to break the loop
+        inNestedResolve = true;
```

Comment:
I do not think having a global flag here is a good idea. As resolving is inherently asynchronous, concurrent resolves can mess up the value of this flag. I fear this problem needs to be solved in Rollup core by correctly tracking uses of the `skipSelf` flag: If a plugin uses `this.resolve` with the `skipSelf` flag, then Rollup needs to remember this when calling `resolveId` of other plugins so that when another plugin subsequently calls `this.resolve` for THE SAME ID (that is important), the original plugin is still skipped (and this should be remembered transitively). I believe this is the issue you are trying to solve?
I thought about this myself some time ago but did not implement anything yet, maybe I should.

Reply:
Ah yep. I was thinking it would only be called once at a time on a single instance, but wasn't thinking about multiple ids in parallel.

>I believe this is the issue you are trying to solve?

I think that would work for this case, although the difference with what you proposed is that the `node-resolve` plugin would be skipped completely when `this.resolve` invoked from commonjs meaning later plugins might get invoked that wouldn't usually, vs now where it would still stop when `node-resolve` returns something (although what it returns now is not really the correct thing, given it could map to an external).

Not sure where the best place to break the chain is, and also when breaking the chain if there is a sensible value that could be returned or if we should actually signal an infinite loop back to the thing that called `this.resolve`.

Confusing problem to think about :)

Reply:
True. But to my understanding, this change is not essential to the functionality you are implementing, right? So you could remove it from here without any regressions and one could always add something later, but it here or in Rollup core.

Reply:
Not quite understanding. The change here to do the nested `this.resolve` is needed (unless there is another solution I'm missing) to fix #609 

I don't mind if the fix is entirely here (with a tweak to keep a `inNestedResolve` flag against the id) or partly here (to do the `this.resolve`) but with a core change to prevent the loops.

Also this isn't a blocker for me btw. I just saw the issue and decided to attempt to fix it, so not sure it's a particularly urgent problem to solve.

Reply:
Ah sorry, should have read the entire thread. The problem is that this is an issue in core that affects all plugins. But to fix it just for node-resolve, you could maintain a data structure that contains the source/importer of what we are just resolving to remember to skip exactly this call when resolving. From Rollup side, each combination will be resolved exactly once. But you could clean up the data structure once a resolution is done to account for plugins resolving the same modules at a later stage.

Reply:
If we later fix this in core, then this fix should not pose a problem as it will just become inactive.

Reply:
A solution from Rollup core would be similar: When skipSelf is used, all nested resolve calls would skip the plugin if the same combination of source and importer is used, but call the plugin for different combinations

Reply:
:+1: I updated the PR to check the `importee` and `importer`.

I'm not sure this is something the core could handle. For this case I _think_ we need to return what the result would be from the nested `resolveId` call, not `null`, because our `this.resolve` is answering the question "If we were to return the result from the node resolve algo, would something else make it external?". If we return `null`, we're not satisfying "If we were to return the result from the node resolve algo".

If we did return `null`

>package-entry-points › uses "require" condition when a module is referenced with require

fails with `CIRCULAR_DEPENDENCY` (in addition to the new test)

Reply:
> I'm not sure this is something the core could handle. For this case I think we need to return what the result would be from the nested resolveId call, not null, because our this.resolve is answering the question "If we were to return the result from the node resolve algo, would something else make it external?". If we return null, we're not satisfying "If we were to return the result from the node resolve algo".

While you are correct that in general that this might cause edge case issues, it would still be correct here as we are calling this.resolve with the resolved id. So the algorithm I have in mind for Rollup core would still pass it to node-resolve once, and even if it would not, node-resolve would not change a resolved id anyway.

I think I will still add this to core eventually because it fixes a lot of fundamental issues with the skipSelf approach (and would only affect secondary `this.resolve` calls that I would assume are usually unlikely to change the actual `resolveId` result, even though this is mostly a heuristic).

Reply:
Maybe I'm not understanding what the algorithm would be

If we have
```
commonjs this.resolve('a') -> node-resolve this.resolve('b') -> commonjs this.resolve('b') -> node-resolve this.resolve('b')
```

(where node-resolve resolves 'a' to 'b')

and that last call returns `null` not `'b'`, that's a problem because commonjs thinks nothing can resolve 'a'.

With your algorithm would that last `this.resolve(b)` just return `'b'` immediately?

Reply:
> that's a problem because commonjs thinks nothing can resolve 'a'

Actually there is always the core resolver. Which would just resolve an absolute URL to itself: https://github.com/rollup/rollup/blob/9bccc4fee3ee19a6f5d252cc4ccbd174b46f6a5d/src/utils/resolveId.ts#L24-L24

There would only be an issue in the browser, but none of those plugins works with the browser build anyway.

Reply:
So in your example, "b" would be an absolute path and the last resolve would again resolve it to "b"

Reply:
Ah got it. There's still the case where if a later plugin does actually resolve to something else that could cause an issue though. Also updating the code to

```ts
    const resolved = await doResolveId(this, importee, importer, opts);
      if (resolved) {
        const resolveContext = { importee: resolved.id, importer };
        // Would the result from the node resoulution algorithm get marked as external by another plugin (or config)?
        // If that's the case we want to mark the import as external to keep the original import intact.
        if (!nestedResolveTracker.inNestedResolve(resolveContext)) {
          const isExternal = await nestedResolveTracker.track(resolveContext, async () => {
            const resolvedResolved = await this.resolve(resolved.id, importer, { skipSelf: true });
            return !!(resolvedResolved && resolvedResolved.external);
          });
          if (isExternal) {
            return false;
          }
        } else {
          // added this
          return null;
        }
```
which I think is going in the direction you're thinking for the core does cause a test failure that has a mix of commonjs and node-resolve.

Reply:
No, it is slightly different. What I am suggesting is to never call `resolveId` with the same arguments twice for any plugin in nested `resolve` calls. In your code, this would be equivalent to

```js
if (resolved) {
  // Would the result from the node resoulution algorithm get marked as external by another plugin (or config)?
  // If that's the case we want to mark the import as external to keep the original import intact.
  if (!nestedResolveTracker.inNestedResolve({ importee, importer })) {
    const isExternal = await nestedResolveTracker.track(
      { importee: resolved.id, importer },
      async () => {
        const resolvedResolved = await this.resolve(resolved.id, importer, {
          skipSelf: true
        });
        return !!(resolvedResolved && resolvedResolved.external);
      }
    );
    if (isExternal) {
      return false;
    }
  } else {
    return null;
  }
}
```

Reply:
So `resolveId` would indeed be called twice, first with the unresolved source and then with the resolved source. Only the third call is skipped when it is called with the resolved source again. This fixes the test in question and only breaks the test that specifically looks for this.
I know it can still cause unexpected situations but I do not think them very likely. But the alternative is that every plugin that uses `this.resolve` needs an elaborate mechanism to guard against this. If there is demand, we could also add another flag to `this.resolve` to allow plugins to specify that they can handle this on their own and do not want the default circularity prevention.

Reply:
I created a preliminary patch in rollup/rollup#4000 to play around with.

Reply:
I have extended the PR in rollup/rollup#4000 to include all the necessary tests and also extended documentation. The idea here would be that if you do not want this behaviour, your "opt-out" is to not use `skipSelf` but entirely implement your own mechanism.
I would be willing to release it if there are no hard concerns against this in a minor release.
Note however that if we want to use this in the node-resolve plugin instead of the custom logic (which I think is a good idea as it reduces complexity), then this would mean changing the peer dependency version of Rollup which would be a breaking change.

Reply:
I'm fine with removing that extra test and simplifying the plugin to use the core change. The edge case is so many layers deep to think about it does seem unlikely, and there's still the way of optiing out if needed in the future like you mentioned in the docs.

It's bugging me though I still can't actually understand why the extra loop round is fixing the test with commonjs 🤔

Reply:
Maybe the question is more why it was broken in the first place without the second round. The reason is that the CommonJS plugin was doing this:

```
this.resolve(importee, importer, {
      skipSelf: true,
      custom: { 'node-resolve': { isRequire: isProxyModule || isRequiredModule } }
    })
```

passing special options to the node-resolve plugin to change resolution, and this is the call that was skipped. Admittedly we could also try to hash the custom options object in the loop prevention, but this is dangerous as we cannot go for object identity as it is likely recreated for each call, and there is no requirement to have it serialisable via JSON stringify.

Reply:
We could do a JSON.stringify in a try-catch as a "best effort" approach.

Reply:
But now I have written this nice description and I am feeling the actual logic is getting more and more complex. Maybe we start with what we have now and see if this causes issues anywhere.

Reply:
Ahhh thanks that was the missing piece. Agree ideally the `custom` should also be part of the key, but waiting until that actually causes a problem (if it does) seems reasonable. Will update this PR